docs/content/posts/2023-04-12-howto-migrate-to-datastream.md

```diff
@@ -0,0 +1,190 @@
+callback.
+
+Another thing: to be able to use Flink state, we need to key the datastream beforeward even if there
+is no group by key because Flink state is designed per-key. Thus, we key by a fake static key so
+that there is a single state.
+
+### [Migrating the limit operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L201-L204)
+
+Like for the order by migration, here also we need to adapt to the streaming semantics in which we
+don't have the whole data at hand. Once again, we use the state API. We apply a mapper that stores
+the incoming rows into
+a [ValueState](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/state/ValueState.html)
+to count them and stop at the set limit. Here also we need to key by a static key to be able to use
+the state API.
+The code resides in
+the [LimitMapper](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L232-L253)
+class.
+
+### [Migrating the sink operation](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkRowDatastream.java#L206-L217)
+
+As with sources, there were big changes in sinks with recent versions of Flink. We now use
+the [Sink interface](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/connector/sink2/Sink.html)
+that requires
+an [Encoder](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/serialization/Encoder.html)
+. But the resulting code is very similar to the one using the DataSet API. It's only
+that [Encoder#encode()](https://javadoc.io/static/org.apache.flink/flink-core/1.16.0/org/apache/flink/api/common/serialization/Encoder.html#encode-IN-java.io.OutputStream-)
+method writes bytes
+when [TextOutputFormat.TextFormatter#format()](https://nightlies.apache.org/flink/flink-docs-release-1.12/api/java/org/apache/flink/api/java/io/TextOutputFormat.TextFormatter.html#format-IN-)
+wrote Strings.
+
+## Conclusion
+
+As you saw for the migration of the join operation, the new unified DataStream API has some
+limitations left in batch mode. In addition, the order by and limit resulting code is quite manual
+and requires the help of the Flink state API for the migration. For all these reasons, the Flink
+community recommends to use Flink SQL for batch pipelines. It results in much simpler code, good
+performance and out-of-the-box analytics capabilities. You could find the equivalent Query3 code
+that uses
+the [Flink SQL/Table API](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/overview/)
+in
+the [Query3ViaFlinkSQLCSV class](https://github.com/echauchot/tpcds-benchmark-flink/blob/master/src/main/java/org/example/tpcds/flink/Query3ViaFlinkSQLCSV.java)
```

Comment:
Let's point to a specific commit instead of the master branch

Reply:
yes, I did that for all the other links in the other articles, thanks for catching this missing one !

Reply:
done